src/Symfony/Component/ObjectMapper/ObjectMapper.php

```diff
@@ -0,0 +1,226 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\ObjectMapper;
+
+use Psr\Container\ContainerInterface;
+use Symfony\Component\ObjectMapper\Attributes\Map;
+use Symfony\Component\ObjectMapper\Exception\RuntimeException;
+use Symfony\Component\PropertyAccess\PropertyAccessorInterface;
+
+/**
+ * Object to object mapper.
+ *
+ * @implements ObjectMapperInterface<T>
+ *
+ * @template T of object
+ *
+ * @experimental
+ *
+ * @author Antoine Bluchet <soyuka@gmail.com>
+ */
+final class ObjectMapper implements ObjectMapperInterface
+{
+    public function __construct(
+        private readonly MapperMetadataFactoryInterface $metadataFactory = new ReflectionMapperMetadataFactory(),
+        private readonly ?PropertyAccessorInterface $propertyAccessor = null,
+        private readonly ?ContainerInterface $serviceLocator = null
+    ) {
+    }
+
+    public function map(object $object, object|string $to = null): object
+    {
+        static $objectMap = null;
+        $objectMapInitialized = false;
+
+        if (null === $objectMap) {
+            $objectMap = new \SplObjectStorage();
+            $objectMapInitialized = true;
+        }
+
+        $refl = new \ReflectionClass($object);
+
+        $metadata = $this->metadataFactory->create($object);
+        $map = $this->getMapTo($metadata, null, $object);
+        $to ??= $map?->to;
+
+        if (!(\is_string($to) && class_exists($to)) && !\is_object($to)) {
+            throw new RuntimeException(sprintf('Attribute of type "%s" expected on "%s.', Map::class, $refl->getName()));
+        }
+
+        $toRefl = new \ReflectionClass($to);
+        $mapped = \is_object($to) ? $to : $toRefl->newInstanceWithoutConstructor();
+        if ($map && $map->transform) {
+            $mapped = $this->applyTransforms($map, $mapped, $mapped);
+        }
+
+        if (!\is_object($mapped)) {
+            throw new RuntimeException('Can not map to a non-object.');
```

Comment:
Can this be a specific exception ? like `BadMapTransformationException` (not sure may be a better name is possible, but i think it should have transform inside since this exception is only possible if a transformer return something else than an object)

Reply:
I really don't understand the gain of adding exceptions like this except adding more classes to maintain.